skbio/metadata/_interval.py

```diff
@@ -0,0 +1,274 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+
+
+class Interval():
+    '''Store the metadata of a sequence interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    _interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
+                 metadata=None, _interval_metadata=None):
+        self._intervals = []
+        for interval in intervals:
+            inv = _polish_interval(interval)
+            self._intervals.append(inv)
+        self.boundaries = boundaries
+        self.metadata = metadata
+        self._interval_metadata = _interval_metadata
+
+    def __getitem__(self, key):
+        return self.metadata[key]
+
+    def __setitem__(self, key, val):
+        self.metadata[key] = val
+
+    def __lt__(self, other):
+        return self.intervals < other.intervals
+
+    def __gt__(self, other):
+        return self.intervals > other.intervals
+
+    def __le__(self, other):
+        return self.intervals <= other.intervals
+
+    def __ge__(self, other):
+        return self.intervals >= other.intervals
+
+    def __eq__(self, other):
+        return ((self.metadata == other.metadata) and
+                (self.intervals == other.intervals) and
+                (self.boundaries == other.boundaries))
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return ''.join([
+            "Interval("
+            "intervals=" + str(self.intervals),
+            ", metadata=" + str(self.metadata),
+            ")"])
+
```

Comment:
That makes sense. I agree, it would be nice to have a `drop` method in the `Interval` object.

I guess my gripe is, is the `Interval` object has may have a reference back to the `IntervalMetadata` object, but isn't actually aware of where exactly it is in the `IntervalMetadata` (i.e. it doesn't know at what position it is stored in `IntervalMetdata._metadata`).

So the following method for drop won't work
```python
    def drop(self):
        self._interval_metadata = None
        self.boundaries = None
        self.intervals = None        
```
Because once you send the references to `None`, now you have a whole bunch of dangling objects. Which is why I originally coded the drop method only in the `IntervalMetadata` object.
The only way I can see incorporating the `drop` method in `Interval` is if the `Interval` object stores the actual index that is it stored within the `IntervalMetadata`.  Then it can update the `_metadata` in `IntervalMetadata`.  I don't think this is a good idea, because this is a lot of book keeping and its a slow linear time update.